torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2331,6 +2437,97 @@ def _rebuild_full_params(self) -> List[Tuple[torch.Tensor, bool]]:
+        ``FULL_SHARD`` and ``SHARD_GRAD_OP``: (1) Being in ``no_sync()`` in the
+        first iteration does not yield a different all-gather sequence, and (2)
+        being in ``no_sync()`` in a later iteration does not give false
+        positive warnings since the all-gather sequence still matches the first
+        iteration sequence (for ``FULL_SHARD``) or the first iteration
+        sequence's prefix (for ``SHARD_GRAD_OP``).
+        """
+        # Only check all-gathers when rebuilding the full parameters in the
+        # forward pass or in the beginning of the backward pass
+        if self.training_state not in (
+            TrainingState_.FORWARD, TrainingState_.BACKWARD_PRE,
+        ):
+            return
+        eod = self._exec_order_data
+        param_index = eod.get_param_index(param)
+        if not eod.is_first_iter:
+            # Only issue warnings on the first deviating iteration and stop
+            # checking thereafter to avoid flooding the console
+            if eod.warn_status == _ExecOrderWarnStatus.WARNED:
+                return
+            # However, we may issue multiple warnings on the first deviating
+            # iteration to help debugging, where either:
+            # 1. This iteration sees more all-gathers than the first iteration
+            msg_prefix = all_gather_seq = None  # non-`None` means we warn
+            if eod.index >= len(eod.param_order):
+                msg_prefix = "Expected no more all-gathers but got an all-gather for "
+                all_gather_seq = eod.param_order + [param_index]
+            else:
+                expected_param_index = eod.param_order[eod.index]
+                # 2. This iteration sees the same number of all-gathers (so
+                # far) but the current parameter to all-gather differs
+                if param_index != expected_param_index:
+                    expected_param_names = eod.get_unflat_param_names(expected_param_index)
+                    assert len(expected_param_names) > 0, \
+                        "The expected parameter to all-gather should always be valid"
+                    msg_prefix = "Expected an all-gather for the FSDP module " \
+                        f"wrapping {expected_param_names} but got an all-gather for "
+                    all_gather_seq = eod.param_order[:eod.index - 1] + [param_index]
+            to_issue_warning = msg_prefix is not None
+            if to_issue_warning:
+                assert all_gather_seq is not None
+                param_names = eod.get_unflat_param_names(param_index)
+                is_added_param = len(param_names) == 0
+                if is_added_param:
+                    msg_suffix = "a newly-added parameter since construction time"
+                else:
+                    msg_suffix = f"the FSDP module wrapping {param_names}"
+                sub_msg = msg_prefix + msg_suffix
+                warnings.warn(
```

Comment:
nit: do we want to warn only once at the first time when the mismatch is found? because the main purpose is for users to notice this change, possibly one warning is sufficient?

Reply:
From earlier offline discussion in the devs chat, I think it is still good to include warnings through the first deviating iteration in case the order change is somehow unintentional. Then, providing some additional information may be helpful for debugging.
I think it should be fine since the messages are still contained to be for one iteration and will not flood the console constantly.